include/mbedtls/sha3.h

```diff
@@ -0,0 +1,262 @@
+
+/**
+ * SHA-3 family id.
+ *
+ * It identifies the family (SHA3-256, SHA3-512, etc.)
+ */
+
+typedef enum
+{
+    MBEDTLS_SHA3_NONE = 0, /*!< Operation not defined. */
+    MBEDTLS_SHA3_224, /*!< SHA3-224 */
+    MBEDTLS_SHA3_256, /*!< SHA3-256 */
+    MBEDTLS_SHA3_384, /*!< SHA3-384 */
+    MBEDTLS_SHA3_512, /*!< SHA3-512 */
+    MBEDTLS_SHA3_SHAKE128, /*!< SHA3-SHAKE128 */
+    MBEDTLS_SHA3_SHAKE256, /*!< SHA3-SHAKE256 */
+    MBEDTLS_SHA3_CSHAKE128, /*!< SHA3-CSHAKE128 */
+    MBEDTLS_SHA3_CSHAKE256, /*!< SHA3-CSHAKE256 */
+} mbedtls_sha3_id;
+
+#if !defined(MBEDTLS_SHA3_ALT)
+// Regular implementation
+//
+
+struct mbedtls_sha3_context;
+typedef struct mbedtls_sha3_family_functions
+{
+    mbedtls_sha3_id id;
+
+    uint16_t r;
+    uint16_t olen;
+    uint8_t xor_byte;
+}
+mbedtls_sha3_family_functions;
+
+/**
+ * \brief          The SHA-3 context structure.
+ *
+ *                 The structure is used SHA-3 checksum calculations.
+ */
+typedef struct mbedtls_sha3_context {
+    uint64_t state[25];
+    uint8_t index;
+    uint8_t id;
+
+    uint16_t r;
+    uint16_t olen;
+    uint8_t xor_byte;
+    uint16_t max_block_size;
+}
+mbedtls_sha3_context;
+
+#else  /* MBEDTLS_SHA3_ALT */
+#include "sha3_alt.h"
+#endif /* MBEDTLS_SHA3_ALT */
+
+/**
+ * \brief          This function initializes a SHA-3 context.
+ *
+ * \param ctx      The SHA-3 context to initialize. This must not be \c NULL.
+ */
+void mbedtls_sha3_init( mbedtls_sha3_context *ctx );
+
+/**
+ * \brief          This function clears a SHA-3 context.
+ *
+ * \param ctx      The SHA-3 context to clear. This may be \c NULL, in which
+ *                 case this function returns immediately. If it is not \c NULL,
+ *                 it must point to an initialized SHA-3 context.
+ */
+void mbedtls_sha3_free( mbedtls_sha3_context *ctx );
+
+/**
+ * \brief          This function clones the state of a SHA-3 context.
+ *
+ * \param dst      The destination context. This must be initialized.
+ * \param src      The context to clone. This must be initialized.
+ */
+void mbedtls_sha3_clone( mbedtls_sha3_context *dst,
+                           const mbedtls_sha3_context *src );
+
+/**
+ * \brief          This function starts a SHA-3 checksum
+ *                 calculation.
+ *
+ * \param ctx      The context to use. This must be initialized.
+ * \param id       The id of the SHA-3 family.
+ *
+ * \return         \c 0 on success.
+ * \return         A negative error code on failure.
+ */
+int mbedtls_sha3_starts( mbedtls_sha3_context *ctx, mbedtls_sha3_id id );
+
+/**
+ * \brief          This function starts a SHA-3 checksum
+ *                 calculation with function-name and customization strings.
+ *
+ * \param ctx      The context to use. This must be initialized.
+ * \param id       The id of the SHA-3 family. It can be \c MBEDTLS_SHA3_CSHAKE128 or
+ *                 \c MBEDTLS_SHA3_CSHAKE256.
+ * \param name     The function-name string. It must be zero-terminated.
```

Comment:
It is unclear from the description if the `_len` parameters should include the NULL termination. Also, the code doesn't appear to actually require the strings to be null terminated.

Reply:
They should not. Initially I did not include the _len parameter and I used `strlen()` but it is a source of problems if the string is not properly formatted. So I added _len parameter. Since sha3 functions accept arbitrary buffers, it also allows non-string parameters. Nevertheless, the FIPS specifies that both parameters shall be strings and being more restrictive/conservative, `name` shall be only from standardized set (only `KMAC` in our case, as `TupleHash` and `ParallelHash` will not be used here).

Reply:
So should we then have

```
\param name     The function-name string.
\param name_len The length of \c name in bytes (not counting any terminating \c NUL character)
```

Note: it's `NUL` not `NULL`. And if `name` doesn't need to be NUL-terminated, don't say that it should be.

Also: should the type of `name` and `custom` be `char *` rather than `uint8_t *`? It's customary in C to use `char *` for "human-readable strings" and `uint8_t *` for "general data"...

Reply:
I think `char *` probably makes sense here, since FIPS does call them strings, although they will be treated as just a sequence of 8-bit values.